### PR TITLE
 [NETBEANS-1696] PHP 7.4 Support

### PR DESCRIPTION
**Pull request by junichi11**
https://issues.apache.org/jira/browse/NETBEANS-1696

## Typed Properties 2.0

See https://wiki.php.net/rfc/typed_properties_v2

- Fix ASTPHP5Parser.cup
- Fix MarkOccurrences, GotoDeclaration, and CodeCompletion features
- Handle unusable types(void and callable) as errors in UnusableTypesUnhandledError class to avoid complicating the grammar(cup file)
- Add and fix unit tests
- Fix FieldRedeclarationHintError (Add the isAnnotation() method to FieldElements)

## Spread Operator in Array Expression

See https://wiki.php.net/rfc/spread_operator_for_array

- Fix ASTPHP5Parser.cup
- Add UnpackableArrayElement as a new AST node (ArrayCreation holds ArrayElements as a list, so UnpackableArrayElement extends not Expression but ArrayElement)
- Fix PHP74UnhandledError
- Improve code completion for const statements (There was no proper context for them, so far, so add GLOBAL_CONST_EXPRESSION and CLASS_CONST_EXPRESSION as context)
- Add unit tests for CC, GotoDeclaration, MarkOccurrences, formatting, and parser

**NOTE** The following cases are not handled in the parser as errors:

- `[$a, ...$b] = [1, 2, 3];`
- `[...42];` `[..."string"];`
- `[...['a' => 'b']];`

We can handle these in a HintError class as errors if needed.

### CODE REVIEW

**Comment by junichi11** [comment]
@tmysik I prefer to create PRs per feature. However, I'm going to add my changes for PHP 7.4 support to this PR(netbeans-1696-php7.4-support branch) because we can't merge changes into the master branch now and the next changes depend on the previous changes. 

New commit is this: https://github.com/apache/netbeans/pull/1314/commits/8af3e52422b1956263acf8259d8502d0ff091f41

**Comment by neilcsmith-net** [comment]
> I prefer to create PRs per feature

You can set up a php7.4 branch in the main repo and set that as the base branch for the PRs, including this one, if that works better for you?  We can then merge from there after NB 11.1.  The release process is not meant to get in the way of using PRs on a per-feature basis.

**Comment by junichi11** [comment]
> You can set up a php7.4 branch in the main repo and set that as the base branch for the PRs

OK, will do. Thanks.

**Comment by junichi11** [comment]
Created the `php7.4` branch, then changed the base branch to it. From now on, I'll create PRs for PHP 7.4 support to the `php7.4` branch.

File: php/php.editor/test/unit/data/testfiles/completion/lib/php74/testSpreadOperatorInArrayExpression/spreadOperatorInArrayExpression.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<?php
+namespace Foo;
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/completion/lib/php74/testSpreadOperatorInArrayExpression/spreadOperatorInArrayExpression.php]: Missing the license header. I'll add it later.

File: php/php.editor/src/org/netbeans/modules/php/editor/completion/PHPCodeCompletion.java [hunk 2]
```diff
@@ -362,6 +370,18 @@ public CodeCompletionResult complete(CodeCompletionContext completionContext) {
                 autoCompleteNamespaces(completionResult, request);
                 autoCompleteExpression(completionResult, request);
                 break;
+            case GLOBAL_CONST_EXPRESSION:
+                autoCompleteNamespaces(completionResult, request);
+                autoCompleteTypeNames(completionResult, request, null, true);
+                autoCompleteConstants(completionResult, request);
+                autoCompleteKeywords(completionResult, request, PHP_GLOBAL_CONST_KEYWORDS);
```
Comment (junichi11) [inline php/php.editor/src/org/netbeans/modules/php/editor/completion/PHPCodeCompletion.java]: Is there anything missing?

Comment (tmysik) [reply php/php.editor/src/org/netbeans/modules/php/editor/completion/PHPCodeCompletion.java]: Not sure now...

Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/completion/PHPCodeCompletion.java]: No problem. Thanks!

**Pull request merged**